src/rust/plugin-work-queue/src/psql_queue.rs

```diff
@@ -0,0 +1,305 @@
+            pipeline_message,
+            &tenant_id,
+            &now,
+            now,
+        )
+        .execute(&self.pool)
+        .await?;
+        Ok(())
+    }
+
+    #[instrument(skip(pipeline_message), err)]
+    pub async fn put_analyzer_message(
+        &self,
+        plugin_id: Uuid,
+        pipeline_message: Vec<u8>,
+        tenant_id: Uuid,
+    ) -> Result<(), PsqlQueueError> {
+        let now = chrono::Utc::now();
+        sqlx::query!(
+            r"
+            INSERT INTO plugin_work_queue.analyzer_plugin_executions (
+                plugin_id,
+                pipeline_message,
+                tenant_id,
+                status,
+                creation_time,
+                last_updated,
+                try_count
+            )
+            VALUES( $1::UUID, $2, $3::UUID, 'enqueued', $4, $5, -1 )
+        ",
+            plugin_id,
+            pipeline_message,
+            &tenant_id,
+            &now,
+            now,
+        )
+        .execute(&self.pool)
+        .await?;
+        Ok(())
+    }
+
+    #[instrument(err)]
+    pub async fn get_generator_message(&self) -> Result<Option<Message>, PsqlQueueError> {
+        // `get_message` does a few things
+        // 1. It attempts to get a message from the queue
+        //      -> Where that message isn't over a day old
+        //      -> Where that message is "visible"
+        //      -> Where that message isn't currently being evaluated by another transaction
+        //      -> Where that message is in the 'enqueued' state
+        // 2. Updates the `try_count`
+        // 3. Updates the `visible_after`
+
+        // Note that:
+        // * messages are invisible for 10 seconds
+        // * messages 'expire' after one day
+        // * messages currently do not have a maximum retry limit
+        // * The one day expiration matches our 1 day partitioning strategy
+
+        // In the future we can leverage a maximum retry limit as well as a batch version of this query
+        // A more dynamic visibility strategy would also be reasonable
+        let request: Option<NextExecutionRequest> = sqlx::query_as!(
+            NextExecutionRequest, r#"
+            UPDATE plugin_work_queue.generator_plugin_executions
+            SET
+                try_count  = plugin_work_queue.generator_plugin_executions.try_count + 1,
+                last_updated = CURRENT_TIMESTAMP,
+                visible_after  = CURRENT_TIMESTAMP + INTERVAL '10 seconds'
+            FROM (
+                 SELECT execution_key, plugin_id, pipeline_message, status, creation_time, visible_after, tenant_id
+                 FROM plugin_work_queue.generator_plugin_executions
+                 WHERE status = 'enqueued'
+                   AND creation_time >= (CURRENT_TIMESTAMP - INTERVAL '1 day')
+                   AND (visible_after IS NULL OR visible_after <= CURRENT_TIMESTAMP)
+                 ORDER BY creation_time ASC
+                 FOR UPDATE SKIP LOCKED
```

Comment:
Wait, you're already using row locks... was my interpretation of visibility wrong?

Reply:
I'm not sure what you're asking. SKIP LOCKED is basically saying "get the next row that can't be immediately locked", so if there's a query currently performing an update a separate query will skip it entirely without waiting.

Here's a post on it:
https://www.2ndquadrant.com/en/blog/what-is-select-skip-locked-for-in-postgresql-9-5/

Reply:
I think we resolved this via slack, right? To reiterate for record, the only locks being used are the ones postgres implicitly acquires when a row is accessed and SKIP LOCKED is a feature where postgres will skip any rows that are currently locked. The visible_after serves as a deadline for the row to be evaluated, and other jobs can pick it up after, without the need to hold any transaction open.

I've added more information on this in the readme.

Reply:
Yeah, I think we resolved this in Slack 👍